sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java

```diff
@@ -351,33 +354,54 @@ public TokenRange(
    */
   protected class WriterImpl<T> implements Writer<T> {
 
+    /**
+     * The threshold of 100 concurrent async queries is a heuristic commonly used
+     * by the Apache Cassandra community. There is no real gain to expect in tuning this value.
+     */
+    private static final int CONCURRENT_ASYNC_QUERIES = 100;
+
     private final CassandraIO.Write<T> spec;
 
     private final Cluster cluster;
     private final Session session;
     private final MappingManager mappingManager;
+    private List<ListenableFuture<Void>> writeFutures;
 
     public WriterImpl(CassandraIO.Write<T> spec) {
       this.spec = spec;
       this.cluster = getCluster(spec.hosts(), spec.port(), spec.username(), spec.password(),
           spec.localDc(), spec.consistencyLevel());
       this.session = cluster.connect(spec.keyspace());
       this.mappingManager = new MappingManager(session);
+      this.writeFutures = Lists.newArrayList();
     }
 
     /**
      * Write the entity to the Cassandra instance, using {@link Mapper} obtained with the
-     * {@link MappingManager}. This method use {@link Mapper#save(Object)} method, which is
-     * synchronous. It means the entity is guaranteed to be reliably committed to Cassandra.
+     * {@link MappingManager}. This method use {@link Mapper#saveAsync(Object)} method, which is
+     * asynchronous. Beam will wait for all futures to complete, to guarantee all writes have
+     * succeeded.
      */
     @Override
-    public void write(T entity) {
+    public void write(T entity) throws ExecutionException, InterruptedException {
       Mapper<T> mapper = (Mapper<T>) mappingManager.mapper(entity.getClass());
-      mapper.save(entity);
+      this.writeFutures.add(mapper.saveAsync(entity));
+      if (this.writeFutures.size() % CONCURRENT_ASYNC_QUERIES == 0) {
```

Comment:
Besides, please know that there will be several instances of `CassandraIO.WriteFn` dofn processing bundles in parallel in the engine (spark/flink ...) cluster. These instances call this write method. So it is possible that you send more than 100 queries to the cassandra cluster even if one dofn has 100 in flight queries and is blocked by the `waitForFuturesToFinish()` until they are executed.

Reply:
1/ you're right, the modulus is useless here and I'll just check the number of elements

2/ It's ok as any cluster should be able to handle 10s/100s of thousands of writes per seconds (if not more). The value of 100 is a heuristic that allows to apply some backpressure per client, and it's expected that the cluster is hammered way more than this through concurrent processes. 
The question is then, if many bundles are being processed at the same time, how can one limit the load on the cluster ? Is there a way to limit concurrency by lowering the number of concurrent bundles or should we make the number of concurrent async queries configurable ?
I'm still inclined to avoid making this configurable as it would bring an additional knob that is specific to Cassandra, and instead deal with concurrency with "standard" Beam knobs to limit job pressure on external systems.

Reply:
you can limit the back pressure at a client level (what you already do) but you can't set the number of bundles. It is the decision of the runner based on its orientation (more batch or more streaming oriented) and the mode (streaming or batch) and also the configuration of the engine (nb of workers for ex). Indeed, it is not exposed to the users as they cannot do better than the runner.
I'm still +1 to not adding knobs for this heuristic value.